Libraries/Opc.Ua.Client/Session.cs

```diff
@@ -5044,13 +5089,39 @@ public IAsyncResult BeginPublish(int timeout)
                 return null;
             }
 
-            SubscriptionAcknowledgementCollection acknowledgementsToSend = null;
+            // get event handler to modify ack list
+            PublishSequenceNumbersToAcknoledgeEventHandler callback = null;
+            lock (m_eventLock)
+            {
+                callback = m_PublishSequenceNumbersToAcknoledge;
+            }
 
             // collect the current set if acknowledgements.
+            SubscriptionAcknowledgementCollection acknowledgementsToSend = null;
             lock (SyncRoot)
             {
-                acknowledgementsToSend = m_acknowledgementsToSend;
-                m_acknowledgementsToSend = new SubscriptionAcknowledgementCollection();
+                if (callback != null)
+                {
+                    try
+                    {
+                        var deferredAcknowledgementsToSend = new SubscriptionAcknowledgementCollection();
+                        callback(this, new PublishSequenceNumbersToAcknoledgeEventArgs(m_acknowledgementsToSend, deferredAcknowledgementsToSend));
+                        acknowledgementsToSend = m_acknowledgementsToSend;
```

Comment:
How about checking that we have not lost an ack after the callback?

Reply:
that could be intentional...